### PR TITLE
Introduce new frames on actions

### LINKED ISSUE #854: The simulator should introduce a new frame on a new action
Consider the following REPL session:

```
>>> .verbosity 3
.verbosity=3
>>> var x: int
>>> action init = x' = 0
>>> action step = x' = x + 1
>>> init.then(step).then(step)
true

 [Frame 0]
 q::input() => true
 ├─ init() => true
 ├─ step() => true
 └─ step() => true

>>>
```

Similar to `quint run`, every invocation of an action should produce its own frame.

### LINKED ISSUE #1133: trace output by `test` disregards the last action
I have a suspicion that `quint test` does not execute the last action. Here is an example:

```
module lastActionInRun {
    var s: str

    action A = 
        s' = "A"
    
    action B = 
        s' = "B"

    
    run t1 = A.then(A).then(B)

    run t2 = A.then(A).then(B).then(s'=s)

    }
```

Upon running `quint test --match=t1 --output=out{#}{}.itf.json lastActionInRun.qnt`, I get two states, `A-A`.

Upon running `quint test --match=t2 --output=out{#}{}.itf.json lastActionInRun.qnt`, I get three states, `A-A-B`

### PR DESCRIPTION
**Pull request by konnov**
Closes #854. This PR fixes the long-standing issue of frames not being introduced in runs when they are executed in REPL. The goal was to fix #1133. However, this would require further refactoring to collect traces in the trace listener. Hence, I have decided to stop here and continue with #1133 once this PR is merged.

There are two major improvements in this PR:

 - [x] Introduction of `onNextState`, which allows us to switch to the next top-level frame:
 
```ts
  /**
   * This callback is called whenever an execution proceeds to the next state,
   * e.g., once A has been evaluated in A.then(B).
   * @param oldState the old state that is about to be discarded
   * @param newState the new state, from which the execution continues
   */
  onNextState(oldState: EvalResult, newState: EvalResult): void
```

 - [x] Introduction of the top-level operator `_`, which normally gets replaced with its child. However, in some cases like evaluating `2 + 3`, we need `_` to produce the verbose output like `_ => 5`.

 - [x] Tests added for any new code
 - [x] Entries added to the respective `CHANGELOG.md` for any new functionality

### CODE REVIEW

File: quint/io-cli-tests.md
Comment (shonfeder) [inline quint/io-cli-tests.md:692]: I don't know how to parse the title of this test.

File: quint/io-cli-tests.md
Comment (shonfeder) [inline quint/io-cli-tests.md:696]: Could you add a note indicating what is meant to be fixed?

File: quint/io-cli-tests.md
Comment (shonfeder) [inline quint/io-cli-tests.md:711]: Since this is not actually testing the output but just the correct construction of the trace, should this perhaps just be in the typescript tests?

In otherwords, given the parsing logic required and what it being tested, I don't understand how/why its a good fit for a black box test.

File: quint/src/runtime/impl/compilerImpl.ts
Comment (shonfeder) [inline quint/src/runtime/impl/compilerImpl.ts:42]: I'm curious, what breaks specifically?

File: quint/src/runtime/impl/compilerImpl.ts
Comment (shonfeder) [inline quint/src/runtime/impl/compilerImpl.ts:266]: I'm not sure how to read this comment. It seems like if `if` branch "does not return" then neither would the `else` branch?

File: quint/src/runtime/trace.ts
Comment (shonfeder) [inline quint/src/runtime/trace.ts:231]: If its an erroneous state as far as we know, should it not raise an error?

File: quint/io-cli-tests.md
Comment (shonfeder) [reply quint/io-cli-tests.md:696]: Seeing how many FIXMEs are in this test: is the test testing anything useful currently? Or is just a sekeleton of a test that should be failing as it, but will be expanded to test something in the future? If the latter, I suggest leaving this test out of the PR.

**Review by shonfeder — APPROVED** [review]
With the caveat that I don't have deep understanding of this code, LGTM, modulo my few questions and minor suggestions.

File: quint/io-cli-tests.md
Comment (konnov) [reply quint/io-cli-tests.md:696]: It's a test I have added when I started working on #1133. Since it is the next issue on my list, I would prefer keeping it in the imperfect shape rather than stashing it somewhere on my disk.

File: quint/io-cli-tests.md
Comment (konnov) [reply quint/io-cli-tests.md:711]: It is testing the actual JSON file, which @ivan-gavran was inspecting by hand

File: quint/src/runtime/impl/compilerImpl.ts
Comment (konnov) [reply quint/src/runtime/impl/compilerImpl.ts:42]: Plenty of tests. I tried to understand, how this was possible, but could not. It really looked like magic to me

File: quint/src/runtime/impl/compilerImpl.ts
Comment (konnov) [reply quint/src/runtime/impl/compilerImpl.ts:266]: Good point. I will fix that.

File: quint/src/runtime/trace.ts
Comment (konnov) [reply quint/src/runtime/trace.ts:231]: in principle, yes, but the trace output is a secondary feature, so I would prefer not to fail on that.

**Pull request merged**
